### ISSUE TITLE
JOSS review

### ISSUE THREAD

**Issue by gcalderone** [p-slash/qsonic#93]
Review for the JOSS submitted paper (original checklist: https://github.com/openjournals/joss-reviews/issues/6373);

The software aims to estimate a non-absorbed mean quasar continuum bluewards of the Lya, and claims to be faster than picca (although providing similar functionalities).

The software appears to be at a rather mature stage, still I believe it would benefit from a **major revision** before the paper is accepted, with changes involving mainly the documentation and the test suite.

## Major changes:

- The software can't be tested on user data unless the user is a member of the DESI collaboration and knows where/how to download the data.  To be more specfic:
  - how can I download the spectra in `desi-y5-1000/spectra-16/`?
  - what is/how can I download the v0 catalog for iron release?
  - what is an imaginary Y5 DESI quickquasars mock?

  Hence I suggest to either declare that the software is targeted exclusively to the members of the DESI collaboration, or to add a complete example involving publicly available data;
  

- The test suite fails with the following errors:
```
======================================================================== short test summary info =========================================================================
FAILED tests/test_calibration.py::test_calibrations - TypeError: NoiseCalibrator.__init__() got an unexpected keyword argument 'add_varlss'
FAILED tests/test_catalog.py::TestCatalog::test_read_validate_adjust_column_names - TypeError: _validate_adjust_column_names() got an unexpected keyword argument 'is_tile'
FAILED tests/test_catalog.py::TestCatalog::test_prime_catalog - TypeError: _prime_catalog() got an unexpected keyword argument 'is_tile'
FAILED tests/test_io.py::TestIOReading::test_read_spectra_onehealpix - AssertionError: assert 2 == 5
FAILED tests/test_mpi_utils.py::TestMPIUtils::test_balance_load - TypeError: balance_load() missing 1 required positional argument: 'mpi_rank'
FAILED tests/test_picca_continuum.py::TestVarLSSFitter::test_add - AttributeError: 'VarLSSFitter' object has no attribute 'calculate_subsampler_stats'. Did you mean: '_calc_subsampler_stats'?
FAILED tests/test_qsonic_fit.py::TestQsonicFit::test_parser - SystemExit: 2
FAILED tests/test_spectrum.py::TestSpectrum::test_simple_coadd - AttributeError: 'Spectrum' object has no attribute 'simple_coadd'
========================================================== 8 failed, 23 passed, 7 skipped, 2 warnings in 8.00s ===========================================================
```

- In the paper, the Author claims that QSOnic is *"faster by an order of magnitude"* w.r.t picca, but no evidence is presented to support the statement.  In particular:
  - Is the improvement mainly due to parallelization? or to a different algorithm?   In the former case, how does performance scale with number of processors?
  - Can a fair comparison be performed?  I.e. can the continuum parametrization of QSOnic (with two diversity parameters) be reduced to the picca one?
  - Are the obtained results compatible with picca?  or are they significantly different (making any performance comparison difficult to interpret) ?
  
  I would suggest to add a complete example using both QSOnic and picca to support the claim.  As an alternative, I suggest to adopt a milder statement such as "may be faster than picca due to the adopted parallelization".

## Minor changes

- The installation instructions seems to be targeted mainly to NERSC users.  I would suggest to move the *Install from PyPi* and *General* sections at the very beginning and to explicitly mention the main dependencies. E.g.
  - is python v3.10 the only accepted version? or is a later version supposed to work as well?
  - is MPI necessary?  If not, how to run the software without it?

- There probably is a typo in the mpi4py installation instructions: it should be `MPICC="mpicc -shared"` rather than `cc`.  If this is not the case it should be clearly stated;

- The "quick start" example shows an `srun` invocation.  Is slurm required to run the software?  If yes it should be clearly stated, otherwise the "quick start" example should not require a job scheduling system;

- In the "How it works" section of the documentation, what do *(1)* and *(2)* means in the first equation?  Are they the extreme of the wavelength range?

- The documentation does not corresponds to the software, e.g. the `get_spectra_reader_function` is not defined despite being mentioned.  Also, a few methods had their name changed , e.g. `read_spectra_onehealpix`;

- In the paper, the Author should clarify which criteria has been used to identify the parallelization domains (sky region / healpix?), as well as which part of the whole process is executed in parallel and which needs to be executed sequentially.

**Comment by p-slash** [comment]
Thank you for the input. I made the changes requested. You can access the updated documentation here: https://qsonic.readthedocs.io/en/paper/index.html

Detailed response below:

> ## Major changes:
> 
>     * The software can't be tested on user data unless the user is a member of the DESI collaboration and knows where/how to download the data.  To be more specfic:
>       
>       * how can I download the spectra in `desi-y5-1000/spectra-16/`?
>       * what is/how can I download the v0 catalog for iron release?
>       * what is an imaginary Y5 DESI quickquasars mock?
>       
>       Hence I suggest to either declare that the software is targeted exclusively to the members of the DESI collaboration, or to add a complete example involving publicly available data;

I added instructions for the publicly available DESI early data. The easiest way to access this data is by having a NERSC account, but the data can be downloaded using Globus, which is noted in the example. Most of the qsonic development has been done after EDR, which introduces some compatibility issues. I provided a workaround that is needed for EDR analysis in the documentation. I also added further examples for DESI members as requested by Corentin Ravoux, and the documentation clearly states that. Relevant pages:

- https://qsonic.readthedocs.io/en/paper/quick_start.html
- https://qsonic.readthedocs.io/en/paper/examples/index.html

>     * The test suite fails with the following errors:
> 
> 
> ```
> ======================================================================== short test summary info =========================================================================
> FAILED tests/test_calibration.py::test_calibrations - TypeError: NoiseCalibrator.__init__() got an unexpected keyword argument 'add_varlss'
> FAILED tests/test_catalog.py::TestCatalog::test_read_validate_adjust_column_names - TypeError: _validate_adjust_column_names() got an unexpected keyword argument 'is_tile'
> FAILED tests/test_catalog.py::TestCatalog::test_prime_catalog - TypeError: _prime_catalog() got an unexpected keyword argument 'is_tile'
> FAILED tests/test_io.py::TestIOReading::test_read_spectra_onehealpix - AssertionError: assert 2 == 5
> FAILED tests/test_mpi_utils.py::TestMPIUtils::test_balance_load - TypeError: balance_load() missing 1 required positional argument: 'mpi_rank'
> FAILED tests/test_picca_continuum.py::TestVarLSSFitter::test_add - AttributeError: 'VarLSSFitter' object has no attribute 'calculate_subsampler_stats'. Did you mean: '_calc_subsampler_stats'?
> FAILED tests/test_qsonic_fit.py::TestQsonicFit::test_parser - SystemExit: 2
> FAILED tests/test_spectrum.py::TestSpectrum::test_simple_coadd - AttributeError: 'Spectrum' object has no attribute 'simple_coadd'
> ========================================================== 8 failed, 23 passed, 7 skipped, 2 warnings in 8.00s ===========================================================
> ```

Maybe the paper branch was broken. I merged main into it. Tests should be fixed. Let me know if any issues remain.

>     * In the paper, the Author claims that QSOnic is _"faster by an order of magnitude"_ w.r.t picca, but no evidence is presented to support the statement.  In particular:
>       
>       * Is the improvement mainly due to parallelization? or to a different algorithm?   In the former case, how does performance scale with number of processors?
>       * Can a fair comparison be performed?  I.e. can the continuum parametrization of QSOnic (with two diversity parameters) be reduced to the picca one?
>       * Are the obtained results compatible with picca?  or are they significantly different (making any performance comparison difficult to interpret) ?
>       
>       I would suggest to add a complete example using both QSOnic and picca to support the claim.  As an alternative, I suggest to adopt a milder statement such as "may be faster than picca due to the adopted parallelization".

We tried improving the speed of picca in various ways. The only remaining possible answer is `multiprocessing` vs `mpi4py`, which is difficult to change in picca. A fair comparison can be performed since the algorithm is the same, and only the implementation details differ. However, I adopted the alternative since the speed improvement is not the crucial part of this package. The paper now says `Due to optimized IO routines, MPI parallelization and efficient numerical algorithms, it may be faster than picca.` in the second to last paragrah.

> 
> ## Minor changes
> 
>     * The installation instructions seems to be targeted mainly to NERSC users.  I would suggest to move the _Install from PyPi_ and _General_ sections at the very beginning and to explicitly mention the main dependencies. E.g.
>       
>       * is python v3.10 the only accepted version? or is a later version supposed to work as well?
>       * is MPI necessary?  If not, how to run the software without it?

Made the changes. I relaxed the python version with a best-practice warning in https://qsonic.readthedocs.io/en/paper/installation.html . MPI is necessary to run qsonic.

>     * There probably is a typo in the mpi4py installation instructions: it should be `MPICC="mpicc -shared"` rather than `cc`.  If this is not the case it should be clearly stated;

Clarified the documentation: ``cc`` command here is a NERSC compiler wrapper that provides optimal compiler options and paths to MPI libraries. Other platforms may use ``mpicc`` instead.

>     * The "quick start" example shows an `srun` invocation.  Is slurm required to run the software?  If yes it should be clearly stated, otherwise the "quick start" example should not require a job scheduling system;

Clarified: You need MPI to run this command, therefore you need to use ``mpirun -np NUMBER_OF_TASKS qsonic-fit ...`` in general or ``srun -n NUMBER_OF_TASKS -c NUMBER_OF_CPUS qsonic-fit ...`` on systems with SLURM.

>     * In the "How it works" section of the documentation, what do _(1)_ and _(2)_ means in the first equation?  Are they the extreme of the wavelength range?

Yes, they are. Added: (1) and (2) superscripts correspond to lower and upper rest-frame wavelength ends considered in the analysis, and are set by ``--forest-w1`` and ``--forest-w2`` respectively.

>     * The documentation does not corresponds to the software, e.g. the `get_spectra_reader_function` is not defined despite being mentioned.  Also, a few methods had their name changed , e.g. `read_spectra_onehealpix`;

Fixed.

>     * In the paper, the Author should clarify which criteria has been used to identify the parallelization domains (sky region / healpix?), as well as which part of the whole process is executed in parallel and which needs to be executed sequentially.

At the end of the first paragraph in the Statement of Need section: Each parallel process is responsible for independent sky regions based on HEALPix [@healpix]. Reading and fitting the continuum are done in parallel. After all quasars are fit with a continuum, all processes syncronize to calculate the mean continuum and check for convergence.

**Comment by gcalderone** [comment]
Thank you!
You addressed all points, and I recommend acceptance after the following **minor changes** are applied:
- In the "Quick start" section you correctly mention the size of the dataset (80 TB).  Please add an indication of how much RAM memory is required to run the job, what is the size of the output and how much time is required to complete the analysis;
- Concering the "Quick start" and "Examples" section:
   - the former is not "quick" since it requires 80TB of data, as well as a pre-processing of the data to solve a problem which is specific to a DESI data release;
   - The latter is only partially an "Examples" section since it relies on the outcome of the "Quick start" section.

  Hence, my suggestion is to move the *"Look into output files"* and *"Simple coadd showcase"* subsections into the *"Quick start"* section.  Also, bring the *"Examples for DESI members"* subsection to the top level (thus eliminating the *"Examples"* section).

- Fix the typo in the paper: syncronize -> synchronize;

Finally, I suggest to prepare and made available online the QSO catalog with the added LASTNIGHT column.  This is only about 25 MB in size and would simplify usage by a newcomer.  However this is a really minor point and is not mandatory for acceptance.

**Comment by p-slash** [comment]
Thanks. I made the changes you asked for. I think they are reasonable.

>   In the "Quick start" section you correctly mention the size of the dataset (80 TB).  Please add an indication of how much RAM memory is required to run the job, what is the size of the output and how much time is required to complete the analysis;
> ...
>   Finally, I suggest to prepare and made available online the QSO catalog with the added LASTNIGHT column. This is only about 25 MB in size and would simplify usage by a newcomer. However this is a really minor point and is not mandatory for acceptance.

I added the following in the ``Workaround for EDR catalogs`` section: This catalog has 38,289 objects, of which ~11,000 are at redshifts z > 2.1 and have their Lya forest region within DESI’s wavelength coverage. You can find an unofficial copy of this catalog here (which is a Google Drive link).

Also the following: This minimal example should calculate quasar continua for about 11,000 spectra. It requires less than 1 GB of RAM and should finish in about a minute for 16 MPI tasks with two cores each. The output files require 320 MB storage. Reading the resolution matrix (by removing --skip-resomat option) multiplies the storage and RAM requirements by a factor of four and slows down reading/writing data. This slow down is neglible for this example, but are relevant for large samples.

>     * Concering the "Quick start" and "Examples" section:
>       
>       * the former is not "quick" since it requires 80TB of data, as well as a pre-processing of the data to solve a problem which is specific to a DESI data release;
>       * The latter is only partially an "Examples" section since it relies on the outcome of the "Quick start" section.
>       
>       Hence, my suggestion is to move the _"Look into output files"_ and _"Simple coadd showcase"_ subsections into the _"Quick start"_ section.  Also, bring the _"Examples for DESI members"_ subsection to the top level (thus eliminating the _"Examples"_ section).

The section is now titled `Tutorial`. The subsections are merged into it. DESI examples have their section.

>     * Fix the typo in the paper: syncronize -> synchronize;

Fixed.

**Comment by gcalderone** [comment]
Looks good to me, thanks!
I close this issue and recommend acceptance of the paper.
